### PR TITLE
nswitch

### LINKED ISSUE #5: a nested version of vswitch, nswitch?
I find `vswitch` is quite useful to me. 

As you stated in the doc, it can be regarded as a special case of `nif()`. It means one of the main advantages besides speed is the conviniences.  I often find a nested version of `vswitch` (maybe we call it `nswitch`?) like below is useful. It's easier to read and write when the values / outputs are not small.

```r
nswitch(
   x,
  "a", 1,
  "b", 2,
  "c", 3,
  default = 4
)
```

However, I have to write it like this for now, due to the missing of such function / API.

```r
vswitch(x, c("a","b","c"), c(1,2,3), default = 4)
```

So I'm opening a FR here to see if we can have such nested version of vswitch or not.

Thanks!

### PR DESCRIPTION
**Pull request by 2005m**
close #5 , @shrektan please let me know what you think. tks

### CODE REVIEW

File: man/vswitch.Rd [hunk 1]
```diff
@@ -1,5 +1,6 @@
-\name{vswitch}
+\name{vswitch/nswitch}
 \alias{vswitch}
+\alias{nswitch}
 \title{ Vectorised switch }
 \description{
   \code{vswitch} is a vectorised version of \code{base} function \code{switch}. This function can also be seen as a particular case of function \code{nif}, as shown in examples below, and should also be faster.
```
Comment (shrektan) [inline man/vswitch.Rd]:
```suggestion
  \code{vswitch}/ \code{nswitch} is a vectorised version of \code{base} function \code{switch}. This function can also be seen as a particular case of function \code{nif}, as shown in examples below, and should also be faster.
```

**Comment by shrektan** [comment]
Great work. Sorry, I was busy for a while. 

(Frankly speaking, I often find myself amazed about your C skills and your familiarity with R internals. :P)

However, I have two questions:

1. Why the two C functions `nswitchR()` and `vswitchR()` can't be merged into just one C function? I think the functionality is basically the same and the difference is just about the R's function signatures.

2. I think the the way of handling encodings is not expected. I mean, the **first** function call should all return `1`. The reason is that when checking encodings, we should compare `x` and `values` under the same UTF-8 encoding. And it looks like currently only `x` is converted to UTF-8.

``` r
library(kit)
#> Attaching kit 0.0.5 (OPENMP enabled using 1 thread)
enc1 = "fa\xE7ile"
Encoding(enc1) = "latin1"
enc2 = enc2utf8(enc1)
nswitch(
  c(enc1, enc2),
  enc1, 1,
  enc2, 2,
  default = 3,
  checkEnc = TRUE
)
#> [1] 2 2
nswitch(
  c(enc1, enc2),
  enc1, 1,
  enc2, 2,
  default = 3,
  checkEnc = FALSE
)
#> [1] 1 2
```

<sup>Created on 2020-11-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by 2005m** [comment]
Answering your questions:
1. Yes you are right but i)`vswitch` is slightly different as it handles list inputs ii) I prefer to isolate each one of them at C level so the code is easier to read and maintain iii) it is easier to deal with R user inputs this way.

2. This is expected because `enc1 == enc2` under the same encoding, and the last value will always be taken. This is actually a difference with `nif` where the first value is taken. To prove you that not only `x` is converted to UTF, you can run this:

```R
enc1 = "fa\xE7ile"
Encoding(enc1) = "latin1"
enc2 = "fa\xE7ileee"
enc2 = enc2utf8(enc2)
nswitch(
  c(enc1, enc2),
  enc1, 1,
  enc2, 2,
  default = 3,
  checkEnc = TRUE
)
```

Please let me know if you do have additional questions. Tks

**Comment by shrektan** [comment]
> This is expected because enc1 == enc2 under the same encoding, and the last value will always be taken. 

Ha, I see. So the logic of handling two equal values is different from `switch()`. Any particular reasons? IMHO, it should have the same behavior as `switch()`:

``` r
x <- "a"
switch(
  x,
  a = 1,
  a = 2
)
#> [1] 1
kit::nswitch(
  x,
  "a", 1,
  "a", 2
)
#> [1] 2
```

<sup>Created on 2020-11-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by 2005m** [comment]
This is to allow execution of the different cases in parallel. In my view the example you provided should not happen when someone is writing code.

**Comment by shrektan** [comment]
The example won't as it uses literal strings. But it's hard to say if the values is from variables. Anyway, I agree the chance should be very very small and we should not worry too much for now, if fixing this affects the performance.

**Pull request merged**